lib/saluki-env/src/autodiscovery/providers/remote_agent.rs

```diff
@@ -0,0 +1,128 @@
+use std::sync::{Arc, Mutex};
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use saluki_error::GenericError;
+use tokio::sync::mpsc::{self, error::TrySendError};
+use tracing::{debug, error, info, warn};
+
+use crate::autodiscovery::{AutoDiscoveryProvider, AutodiscoveryEvent, Config};
+use crate::helpers::remote_agent::RemoteAgentClient;
+
+/// An autodiscovery provider that uses the Datadog Agent's internal gRPC API to receive autodiscovery updates.
+pub struct RemoteAgentAutoDiscoveryProvider {
+    client: RemoteAgentClient,
+    event_senders: Arc<Mutex<Vec<mpsc::Sender<AutodiscoveryEvent>>>>,
+    listener_started: bool,
+}
+
+impl RemoteAgentAutoDiscoveryProvider {
+    /// Creates a new `RemoteAgentAutoDiscoveryProvider` that uses the remote client to receive autodiscovery updates.
+    pub fn new(client: RemoteAgentClient) -> Self {
+        Self {
+            client,
+            event_senders: Arc::new(Mutex::new(Vec::new())),
+            listener_started: false,
+        }
+    }
+
+    async fn start_background_listener(&mut self) -> Result<(), GenericError> {
+        if self.event_senders.lock().unwrap().is_empty() {
+            return Err(GenericError::msg(
+                "Cannot start background listener without subscribers",
+            ));
+        }
```

Comment:
Since the call this function after pushing a sender, we can remove this check and avoid acquire the lock twice on the first `subscribe` call